### ISSUE TITLE
error calling EditFacts for multiple facts

### ISSUE THREAD

**Issue by ghost** [28msec/cellstore-csharp#63]
Hi,
I am trying to edit multiple facts together using this code: 
`dataAPI.EditFacts(token: rwtoken, patch: patch, aid: aid, dimensions: dimensions);`

If patch type is List of Fact class i recive the error message listed below.
If patch type is Fact the operation is successfull.

Error calling EditFacts: { "request_id" : "57b45337d38b6126b73b07ae", "context" : "boi.local.28.io", "type" : "type", "code" : "XPTY0004", "location" : { "module" : "/public/api/facts/PATCH.jq", "line-number" : 69, "line-number-end" : 69, "column-number" : 5, "column-number-end" : 86 }, "description" : "array() can not be treated as type object()_", "message" : "</public/api/facts/PATCH.jq>:69,5: type error [err:XPTY0004]: array() can not be treated as type object()_" }

Fact class is:
public class Fact
    {
        public List<string> KeyAspects { get; set; }
        public Dictionary<string, string> Aspects { get; set; }
        public Profiles Profiles { get; set; }
        public string Value { get; set; }
        public int Decimals { get; set; }
        public List<AuditTrails> AuditTrails { get; set; }
    }

**Comment by fcavalieri** [comment]
Hello @amit-boi , thanks for the report.

In the EditFacts method, the patch can only be a single object. That is what the not so friendly error is trying to tell you. The patch object is merged into each identified fact as a result of the method call so it must be one. In the C# bindings bodies are defined as object and toString is invoked as necessary.

What is the intended behaviour when you pass multiple patch objects?

Kind regards,
Federico

**Comment by ghost** [comment]
I want to update all the objects in one call in order to improve performance.
What is the recommanded way to update high number of facts ?

**Comment by fcavalieri** [comment]
Hello @amit-boi,

to update more than one fact at the same time you don't need to pass multiple patches.
Let's consider a case, we have an archive with aid "open_table" and mark all its invalid facts as valid.

To do so we can:
PATCH http://boi-demo.28.io/v1/_queries/public/api/facts?token=barfoo&xbrl28:Archive=open_table&open=true&boi:FilingStatus=invalid

Body:
{
  "Aspects": {
    "boi:FilingStatus": "valid"
  }
}

This part of the request identifies the facts (all the facts from open_table that are marked as invalid): xbrl28:Archive=open_table&open=true&boi:FilingStatus=invalid
To each of them we set Aspects.boi:FilingStatus to valid.

Are you trying to patch each object with a different patch?

Let me know if this helps,
Federico

**Comment by ghost** [comment]
Yes.
I am changing the value of every object so i need diffrent patch for each object.

**Comment by fcavalieri** [comment]
Hello @amit-boi,
could you make me an example with two facts, two patches and two expected obtained facts? I am trying to fully understand the use case because changing the value of every object does not necessarily imply needing different patches. 

The behaviour of the method is:
foreach fact x that matches query parameters
"patch" fact x with the given patch

Thanks and kind regards,
Federico

**Comment by ghost** [comment]
If the value of every patch is diffrent and it's for diffrent fact then i think i need diffrent patches.

Anyway, I am trying to do editFacts in parallel threads for high number of facts and is seems to solve the performance issue.

**Comment by fcavalieri** [comment]
If you want to apply different patches to different facts you need to make multiple requests at the moment. If you pass multiple patches in the same call on every identified fact all patches are applied in order. 
(This is not working in your code because the serialization of lists of objects in the body of a request is handled differently.

**Comment by ghost** [comment]
i did not understand your second sentence. can you show me an example ?
